### ISSUE TITLE
Auto tab-complete paths and files as arguments to `echo`

### ISSUE THREAD

**Issue by casey** [fish-shell/fish-shell#6856]
I would like to tab-complete paths and files that are arguments to `echo`. I find find this helpful, for example, when I want to insert the name of a file into a config file:

```
: echo path/to/some/file >> some/config/file
```

Or when I want to get a path into the clipboard:

```
: echo ~/.ssh/id_ed25519.pub | pbcopy
```

It appears though that paths will not tab-complete when they are arguments to `echo`:

https://asciinema.org/a/JzHgNe4dIwsZHnjbm7aaOT2Vi

I tried creating `~/.config/fish/completions/echo.fish`, in an attempt to suppress special completion behavior for echo, but to no avail. I suppose this is because it is a built-in function.

Ideally, it would be nice to be able to add an option to always auto-complete paths, regardless of the context.

Shell and system information:

```
: fish --version
fish, version 3.1.0
: echo $version
3.1.0
: uname -a
Darwin durandal 18.7.0 Darwin Kernel Version 18.7.0: Tue Aug 20 16:57:14 PDT 2019; root:xnu-4903.271.2~2/RELEASE_X86_64 x86_64
: echo $TERM
screen
```

Running `sh -c 'env HOME=$(mktemp -d) fish'` did not change the behavior I am reporting.

**Comment by faho** [comment]
Yeah, we suppress file completions for most builtins.

The workaround is to erase the completions for echo - `complete -c echo --erase`.

We probably should favor completing files here as opposed to _nothing_.

**Comment by casey** [comment]
Thanks for the workaround!

If I run `complete -c echo --erase`, I can start completing paths to `echo`, but if add the same command to `~/.config/fish/config.fish`, it doesn't seem to work. Is there somewhere else I should put it to make the configuration permanent?

**Comment by faho** [comment]
>Is there somewhere else I should put it to make the configuration permanent?

This is done *after* config.fish, when we do the interactive setup, right before the first prompt.

So technically any event handler (fish_preexec, fish_postexec, fish_prompt) would work.

Or make a function that just wraps echo, but don't call it `echo` or set up wrapping completions for it:

```fish
function e
    echo $argv
end
```

**Comment by casey** [comment]
Is there and event handler that runs once after interactive setup? It feels a bit profligate to put it in fish_preexec, fish_postexec, or fish_prompt.

**Comment by krobelus** [comment]
> Is there and event handler that runs once after interactive setup? 

You can define this function in your `config.fish` that runs at the first prompt, which is after the builtin completions are defined.

```fish
function fixup-builtin-completions --on-event fish_prompt
    complete -c echo --erase
    functions --erase fixup-builtin-completions
end
```

> We probably should favor completing files here as opposed to nothing.

I agree, I think generic file completion could potentially be useful for `contains`, `count`, `echo`, `printf`, `switch`, and `string`.

**Comment by casey** [comment]
@krobelus Great, that works. Thank you!

I would favor the ability to enable generic file completion everywhere, but, barring that, I think generic file completion is useful in any context where a string is expected is probably useful.

Although it may be possible to enumerate the contexts where file completion is useful, inevitably, there will be omissions, which will may be jarring to the end user when some contexts are unexpectedly inconsistent.

**Comment by ridiculousfish** [comment]
I think file completions are usually better than nothing, except when the command requires a subcommand (`git <tab>`, `string <tab>`, `if <tab>`, etc.). Allowing file completions on builtins seems like a good thing to add with or after #6854.

**Comment by faho** [comment]
Fixed in ff68bdcebab76c288f9d672abe246c837904251b

(had the wrong tab open)
